go/service/saltpack.go

```diff
@@ -269,6 +241,251 @@ func (h *SaltpackHandler) SaltpackVerifyString(ctx context.Context, arg keybase1
+	if opts.Signed {
+		auth = keybase1.AuthenticityType_SIGNED
+	}
+	return keybase1.SaltpackEncryptOptions{
+		Recipients:       opts.Recipients,
+		AuthenticityType: auth,
+		NoSelfEncrypt:    !opts.IncludeSelf,
+		UseEntityKeys:    true,
+	}
+}
+
+func (h *SaltpackHandler) frontendEncrypt(ctx context.Context, sessionID int, arg *engine.SaltpackEncryptArg) error {
+	uis := libkb.UIs{
+		SecretUI:  &nopSecretUI{},
+		SessionID: sessionID,
+	}
+
+	keyfinderHook := saltpackkeys.NewRecipientKeyfinderEngineHook(false)
+
+	eng := engine.NewSaltpackEncrypt(arg, keyfinderHook)
+	m := libkb.NewMetaContext(ctx, h.G()).WithUIs(uis)
+
+	return engine.RunEngine2(m, eng)
+}
+
+func (h *SaltpackHandler) frontendDecrypt(ctx context.Context, sessionID int, arg *engine.SaltpackDecryptArg) (keybase1.SaltpackEncryptedMessageInfo, bool, error) {
+	spui := &capSaltpackUI{}
+	uis := libkb.UIs{
+		IdentifyUI: h.NewRemoteIdentifyUI(sessionID, h.G()),
+		SecretUI:   &nopSecretUI{},
+		SaltpackUI: spui,
+		SessionID:  sessionID,
+	}
+	m := libkb.NewMetaContext(ctx, h.G()).WithUIs(uis)
+	resolver := saltpackkeys.NewKeyPseudonymResolver(m)
+	eng := engine.NewSaltpackDecrypt(arg, resolver)
+	if err := engine.RunEngine2(m, eng); err != nil {
+		return keybase1.SaltpackEncryptedMessageInfo{}, false, err
+	}
+	return eng.MessageInfo(), spui.verified, nil
+}
+
+func (h *SaltpackHandler) frontendVerify(ctx context.Context, sessionID int, arg *engine.SaltpackVerifyArg) (*capSaltpackUI, error) {
+	spui := &capSaltpackUI{}
+	uis := libkb.UIs{
+		IdentifyUI: h.NewRemoteIdentifyUI(sessionID, h.G()),
+		SecretUI:   &nopSecretUI{},
+		SaltpackUI: spui,
+		SessionID:  sessionID,
+	}
+	eng := engine.NewSaltpackVerify(h.G(), arg)
+	m := libkb.NewMetaContext(ctx, h.G()).WithUIs(uis)
+	if err := engine.RunEngine2(m, eng); err != nil {
+		return nil, err
+	}
+	return spui, nil
+}
+
+func (h *SaltpackHandler) frontendSign(ctx context.Context, sessionID int, arg *engine.SaltpackSignArg) error {
+	uis := libkb.UIs{
+		SecretUI:  &nopSecretUI{},
+		SessionID: sessionID,
+	}
+	m := libkb.NewMetaContext(ctx, h.G()).WithUIs(uis)
+	eng := engine.NewSaltpackSign(h.G(), arg)
+	return engine.RunEngine2(m, eng)
+}
+
+func boxFilename(inFilename, suffix string) (string, *libkb.BufferWriter, error) {
+	dir, file := filepath.Split(inFilename)
+	withExt := filepath.Join(dir, file+suffix+saltpackExtension)
+	f, err := os.Create(withExt)
+	if err != nil {
+		return "", nil, err
+	}
+	return withExt, libkb.NewBufferWriter(f), nil
+}
+
+func unboxFilename(inFilename, suffix string) (string, *libkb.BufferWriter, error) {
+	dir, file := filepath.Split(inFilename)
+	res := file + suffix
+	if strings.HasSuffix(file, encryptedExtension) {
+		res = strings.TrimSuffix(file, encryptedExtension)
+	} else if strings.HasSuffix(file, signedExtension) {
+		res = strings.TrimSuffix(file, signedExtension)
+	}
+	out := filepath.Join(dir, res)
+
+	var found bool
+	for i := 0; i < 100; i++ {
+		poss := out
```

Comment:
this whole method is unfortunately kinda weird. might be nice if some of these variables in here had slightly more descriptive names.